### DISCUSSION TITLE
发商城表情的时候 MessageChain 长度只有 1

### DISCUSSION THREAD

**Discussion by Mystery406** [mamoe/mirai#815] [Q&A] [answered]
/**
     * 消息内容.
     *
     * 第一个元素一定为 [MessageSource], 存储此消息的发送人, 发送时间, 收信人, 消息 ids 等数据.
     * 随后的元素为拥有顺序的真实消息内容.
     */
    public override val message: MessageChain

看doc意思好像是收到的MessageChain长度应该>1,为啥发商城表情的时候MessageChain长度只有1,后面就没了

**Comment by Him188** [comment, marked as answer]
发送出去的消息的 MessageChain 是你构造的，元素由你控制。从服务器接收的消息由 mirai 控制，因此会添加 MessageSource 到第一个元素方便操作。

如果你希望获取消息回执用于引用或撤回你发出去的消息，参考 `MessageSource`

**Reply by Mystery406** [reply]
我是希望别人单发给我一个商城表情后能从MessageChain的第二个元素获取到，和别人单发一个At一样。

**Reply by Him188** [reply]
所以你收到的实际消息是什么

**Reply by Him188** [reply]
@sandtechnology

**Reply by sandtechnology** [reply]
怎么了（
我解析的时候确实没啥问题啊，但是商城表情是单个的信息，所以和闪照一样继承了ConstrainSingle，其他没有区别了啊（

**Reply by Mystery406** [reply]
val s = message.contentToString()
这个message是MarketFaceImpl的时候s=""

**Reply by Him188** [reply]
请提供完整代码

**Reply by Mystery406** [reply]
不知道咋上传图片...
```kotlin
eventChannel.subscribeAlways<MessageEvent> {event ->
    val messageContent = buildString {
        for (msg in event.message) {
            append(
                when (msg) {
                    is Image -> "[image:${msg.queryUrl()}]"
                    is FlashImage -> "[flash:${msg.image.queryUrl()}]"
                    is At -> "[at:${msg.target}]"
                    else -> {
                        val s = msg.contentToString()
                        s
                    }
                }
            )
        }
    }
}
```
msg为MarketFaceImpl的时候，s为“”。
应该是MarketFace的contentToString=name ,name是“”导致的

**Reply by sandtechnology** [reply]
我记得这种情况是客户端行车不规范 请使用is MarketFace判断（

**Reply by Mystery406** [reply]
> 
> 
> 我记得这种情况是客户端行车不规范 请使用is MarketFace判断（

用is MarketFace判断也是空，应该是contentToString=name的时候，name是空导致的

**Reply by sandtechnology** [reply]
那你客户端发的是空的我有啥办法嘛（

**Reply by sandtechnology** [reply]
哦 这种情况下末尾的兼容信息是对的 确实需要发个PR修一下了（
你可以暂时通过末尾的兼容信息判断名称（如果为空）

**Reply by sandtechnology** [reply]
已经发送PR，请等待修复（
